gammapy/utils/fitting/sampling.py

```diff
@@ -0,0 +1,162 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# helper functions for mcmc smapling
+import numpy as np
+import emcee
+import matplotlib.pyplot as plt
+import corner
+
+__all__ = ["uniform_prior", "run_mcmc", "plot_trace", "plot_corner"]
+
+#TODO: so far only works with a uniform prior on parameters
+
+# Prior functions
+def uniform_prior(value, umin, umax):
+    """Uniform prior distribution."""
+    if umin <= value <= umax:
+        return 0.0
+    else:
+        return -np.inf
+
+
+def normal_prior(value, mean, sigma):
+    """Normal prior distribution."""
+    return -0.5 * (2 * np.pi * sigma) - (value - mean) ** 2 / (2.0 * sigma)
+
+
+# Read/write parameters in the dataset
+def model_to_par(dataset):
+    """
+    Return a tuple of the factor parameters of all
+    free parameters in the dataset sky model.
+    """
+    pars = []
+    for p in dataset.parameters.free_parameters:
+        pars.append(p.factor)
+
+    return pars
+
+
+def par_to_model(dataset, pars):
+    """Update model in dataset with a list of free parameters factors"""
+    for i, p in enumerate(dataset.parameters.free_parameters):
+        p.factor = pars[i]
+
+
+# Compute LogLike associated with prior and data/model evaluation
+def lnprior(dataset):
+    """
+    Return probability of parameter values according to prior knowledge.
+    Parameter limits should be done here through uniform prior ditributions
+    """
+    logprob = 0
+    for par in dataset.parameters.free_parameters:
+        logprob += uniform_prior(par.value, par.min, par.max)
+
+    return logprob
+
+
+def lnprob(pars, dataset):
+    """Estimate the likelihood of a model including prior on parameters."""
+    # Update model parameters factors inplace
+    for factor, par in zip(pars, dataset.parameters.free_parameters):
+        par.factor = factor
+
+    lnprob_priors = lnprior(dataset)
+
+    # dataset.likelihood returns Cash statistics values
+    # emcee will maximisise the LogLikelihood so we need -dataset.likelihood
+    total_lnprob = -dataset.likelihood() + lnprob_priors
+
+    return total_lnprob
+
+
+def run_mcmc(dataset, nwalkers=8, nrun=1000, threads=1):
```

Comment:
Docstring missing `threads` in parameters list, and `Returns`.

Reply:
Done